source/event_loop.c

```diff
@@ -9,45 +9,115 @@
 #include <aws/common/system_info.h>
 #include <aws/common/thread.h>
 
-int aws_event_loop_group_init(
-    struct aws_event_loop_group *el_group,
+static void s_event_loop_group_thread_exit(void *user_data) {
+    struct aws_event_loop_group *el_group = user_data;
+
+    aws_simple_completion_callback *completion_callback = el_group->shutdown_options.shutdown_callback_fn;
+    void *completion_user_data = el_group->shutdown_options.shutdown_callback_user_data;
+
+    aws_mem_release(el_group->allocator, el_group);
+
+    if (completion_callback != NULL) {
+        completion_callback(completion_user_data);
+    }
+
+    aws_global_thread_tracker_decrement();
+}
+
+static void s_aws_event_loop_group_shutdown_sync(struct aws_event_loop_group *el_group) {
+    while (aws_array_list_length(&el_group->event_loops) > 0) {
+        struct aws_event_loop *loop = NULL;
+
+        if (!aws_array_list_back(&el_group->event_loops, &loop)) {
+            aws_event_loop_destroy(loop);
+        }
+
+        aws_array_list_pop_back(&el_group->event_loops);
+    }
+
+    aws_array_list_clean_up(&el_group->event_loops);
+}
+
+static void s_event_loop_destroy_async_thread_fn(void *thread_data) {
+    struct aws_event_loop_group *el_group = thread_data;
+
+    s_aws_event_loop_group_shutdown_sync(el_group);
+
+    aws_thread_current_at_exit(s_event_loop_group_thread_exit, el_group);
+}
+
+static void s_aws_event_loop_group_shutdown_async(struct aws_event_loop_group *el_group) {
+
+    struct aws_thread cleanup_thread;
+    AWS_ZERO_STRUCT(cleanup_thread);
+
+    AWS_FATAL_ASSERT(aws_thread_init(&cleanup_thread, el_group->allocator) == AWS_OP_SUCCESS);
+
+    struct aws_thread_options thread_options;
+    AWS_ZERO_STRUCT(thread_options);
+
+    AWS_FATAL_ASSERT(
+        aws_thread_launch(&cleanup_thread, s_event_loop_destroy_async_thread_fn, el_group, &thread_options) ==
```

Comment:
I've got a way to avoid these FATAL_ASSERTS, but it's probably too much tinkering to do right now when we really want this in mainline. Maybe just put in a TODO:

/* TODO: get rid of FATAL_ASSERTS by initializing and launching cleanup thread during event-loop-group initialization. Cleanup thread waits until it's signaled via condition variable. */